[API Compatiblity] Sum c sink new

zhengshengning opened the pull request:
<!-- TemplateReference: https://github.com/PaddlePaddle/Paddle/wiki/PULL-REQUEST-TEMPLATE--REFERENCE -->
<!-- Demo: https://github.com/PaddlePaddle/Paddle/pull/24810 -->

### PR Category
<!-- One of [ User Experience | Execute Infrastructure | Operator Mechanism | CINN | Custom Device | Performance Optimization | Distributed Strategy | Parameter Server | Communication Library | Auto Parallel | Inference | Environment Adaptation ] -->
Operator Mechanism

### PR Types
<!-- One of [ New features | Bug fixes | Improvements | Performance | BC Breaking | Deprecations | Docs | Devs | Not User Facing | Security | Others ] -->
Improvements

### Description
<!-- Describe what you’ve done -->

下沉 paddle.sum 和 paddle.Tensor.sum 到C++层：

1. 使用args mapper解决python api和_C_ops 中参数顺序不一致的情况。
2. 使用pre_process进行前处理。

**修改说明：**
1. 修复：当arg_map和pre_process同时存在时，代码生成PreProcess 不出现在 static_op_function.cc中的问题。
2. 单测修改：test/dygraph_to_static/test_error.py 中 NpApiErr()调用的np.sum()，替换成np.count_nonzero()。修改原因：因为之前 sum 不支持 out 参数所以会抛出异常，但是本pr下沉sum后，支持了out参数，所以在静态图下不会抛出异常。 
pcard-67164

zhengshengning committed:
add sum sink

zhengshengning committed:
Merge branch 'develop' of https://github.com/PaddlePaddle/Paddle into sum_c_sink_new

zhengshengning committed:
Merge branch 'develop' of https://github.com/PaddlePaddle/Paddle into sum_c_sink_new

zhengshengning committed:
rm old *_depreeated

SigureMo commented on test/prim/pir_prim/test_vjp_prim.py:
```diff
@@ -46,27 +46,6 @@ def get_ir_divide_program():
-            dout.stop_gradient = False
-            out = paddle.sum(x)
-        pir_program = pir.translate_to_pir(main_program.desc)
-        return pir_program
```
不要删，改成纯 PIR 组网即可

```python
def get_ir_divide_program():
    paddle.enable_static()
    main_program, start_program = (
        paddle.static.Program(),
        paddle.static.Program(),
    )
    with paddle.static.program_guard(main_program, start_program):
        x = paddle.tensor.fill_constant(
            shape=[1, 4], dtype='float32', value=2.0
        )
        x.stop_gradient = False
        y = paddle.tensor.fill_constant(shape=[4], dtype='float32', value=1.0)
        y.stop_gradient = False
        dout = paddle.tensor.fill_constant(
            shape=[1, 4], dtype='float32', value=1.0
        )
        dout.stop_gradient = False
        out = paddle.divide(x, y)

    return main_program


def get_ir_sum_program():
    paddle.enable_static()
    main_program, start_program = (
        paddle.static.Program(),
        paddle.static.Program(),
    )
    with paddle.static.program_guard(main_program, start_program):
        x = paddle.tensor.fill_constant(
            shape=[4, 5], dtype='float32', value=2.0
        )
        x.stop_gradient = False
        dout = paddle.tensor.fill_constant(shape=[], dtype='float32', value=1.0)
        dout.stop_gradient = False
        out = paddle.sum(x)
    return main_program
```

divide 和 sum 一起改一下

SigureMo commented on test/dygraph_to_static/test_error.py:
```diff
@@ -353,7 +353,7 @@ def test_key_error(self):
 def NpApiErr():
     a = paddle.to_tensor([1, 2])
-    b = np.sum(a.numpy())
+    b = np.count_nonzero(a.numpy())
```
补充说明：

本单测用于测试动转静能够正确捕获 numpy error，但是本 PR 为 `paddle.sum/paddle.Tensor.sum` 添加了参数 out，使得这里不再报错，因为 `np.sum` 的 dispatch 逻辑如下：

<img width="686" height="341" alt="Image" src="https://github.com/user-attachments/assets/2ab422df-febe-4ceb-a83d-707741760bb2" />

即对于非 np 数据 x，`np.sum(x)` 会转发到 `x.sum()`，但是会传递 kwarg out，而此前 sum 是不支持的，因此报错

本 PR 支持后不再报错，因此调整 API 为 `np.count_nonzero`

SigureMo commented on test/ir/pir/test_ir_backward.py:
```diff
@@ -216,40 +216,6 @@ def test_concat(self):
-            x_s.stop_gradient = False
-            k_s = paddle.sum(x_s, axis=(-1,), keepdim=False)
-        pir_program = pir.translate_to_pir(main_program.desc)
-        return pir_program
```
和 `test_vjp_prim` 单测一样，改写为纯 PIR case

```diff
def get_ir_program():
-    with paddle.pir_utils.OldIrGuard():
-        with_block()
+    with_block()
-    pir_program = pir.translate_to_pir(main_program.desc)
-    return pir_program
+    return main_program
```

这个文件的 get_ir_program_1、2、3 一起改一下

SigureMo replied:
可参考 #74962，其他文件我单独提 PR 统一修改了下，这两个文件你在这个 PR 改一下吧

zhengshengning replied:
好嘞！

SigureMo commented on test/legacy_test/test_instance_norm_op_v2.py:
```diff
@@ -206,7 +206,7 @@ def setUp(self):
         self.python_api = instance_norm_wrapper
         self.public_python_api = instance_norm_wrapper
         self.check_prim = (
```
直接 `self.check_prim = False`

SigureMo replied:
下同

SigureMo reviewed (commented):
这俩是测 PIR 功能的，只是早期 PIR 组网没建设完全因此先老 IR 组网后利用了 PT（program translaotr）转换为 PIR program，现在应该改成纯 PIR 组网的 case 而不是直接删掉，改法都给出示例了，我这边改完直接就可以跑过（develop 分支、mac 本地）

zhengshengning committed:
Merge branch 'develop' of https://github.com/PaddlePaddle/Paddle into sum_c_sink_new

zhengshengning committed:
fix test case

zhengshengning committed:
add # type: ignore

zhengshengning committed:
fix

zhengshengning committed:
merge develop

zhengshengning committed:
Merge branch 'develop' of https://github.com/PaddlePaddle/Paddle into sum_c_sink_new

zhengshengning committed:
fix

zhengshengning committed:
fix

zhengshengning committed:
fix

zhengshengning committed:
fix1

zhengshengning committed:
fix2

zhengshengning commented:
/re-run all-failed

zhengshengning committed:
merge develop

co63oc committed:
replace mkldnn_data_type in test_pool2d_bf16_onednn_op (#75046)

DanielSun11 committed:
[API compatibility]sinking log2 into c++

YqGe585 committed:
missing kernel log in static mode (#75021)

D0m021ng committed:
[Feat]: add LibUVTCPServer backend for TCPStore (#74389) (#74975)

* [Feat]: add LibUVTCPServer backend for TCPStore (#74389)

* TCPStore adds LibUVTCPServer backend

* update parameters

* fix ut

* update

* update log

* update cmake

* update reinterpret_cast

* fix ut

* update tcp_store_using_libuv default value

* remove some cmake

* add libuv win build

* code format

* fix win32 deps

* fix win32 build

* update

* update

* update libuv cmake

* add header file

* update

* update

* format code

* format code

* update libuv cmake

* format code

* format code

* update cmake

* add libuv deps on macos

* update libuv cmake

* update libuv cmake

* update

* fix conflicts

* update

* fix

* fix libucrt.lib

* fix

* update

* update libuv.cmake

* fix libuv cmake

* fix

* fix libuv.cmake

fxyfxy777 committed:
new cuda api  compat with torch(#75063)

xiaoguoguo626807 committed:
【FlexCP】add Skip param param for merge_shard_state_dict (#75061)

* fix data is nullptr

* add dist merge

* change test

* change test

* add skip optimizer param

lshpku committed:
[DeepEP] Remove num_nvl_bytes limit under low-latency mode (#75058)

co63oc committed:
rename test_log_softmax_mkldnn_op (#74728)

DanielSun11 committed:
Fix args mapper mechanism  for sinking  Python API into C++ (#75008)

* fix arg mapper

* sink argmax argmin

* fix

Enigmatisms committed:
[PHI] PyTorch aligned scatter reduce kernels for integer mean (#75039)

* [PHI] PyTorch aligned scatter_reduce_mean

Note: change of behavior for integer type scatter reduce mean.
PyTorch: div floor (also for negatives), original paddle: round to 0.

* [PHI] Add CPU integer floor div

* [PHI] Exclude XPU for mean int testing

Add CPU testing so that there is no need to skip coverage

* notest,test=coverage

SigureMo committed:
[Compat] Refine torch library compat code (#75012)

DongBaiYue committed:
[API compatibility] paddle.unbind (#75056)

zeroRains committed:
fix the bug in isinstance(arg, int) failed (#75064)
